### DISCUSSION TITLE
Declarative shadow dom in a server component?

### DISCUSSION THREAD

**Discussion by TychosDrunkElk** [vercel/next.js#51717] [App Router]
I have an app using [declarative shadow dom](https://developer.chrome.com/en/articles/declarative-shadow-dom/) to render a header and footer fetched from an api. Ideally, I could make a server component like so:

```jsx
import { headers } from 'next/headers';

export async function ChromeSSR({ children }) {
  const nextHeaders = headers();
  const response = await fetch(api, {headers: nextHeaders});
  const html = await response.text();
  return (
    <div id="__chrome" style={{ display: 'contents' }} >
      <template
        shadowrootmode="open"
        suppressHydrationWarning
        dangerouslySetInnerHTML={{
          __html: html,
        }}
      />
      {/* https://web.dev/declarative-shadow-dom/#polyfill goes here*/}
      {children}
    </div>
  );
}

```
but this will always give me the error `Warning: Expected server HTML to contain a matching <template> in <div>.`
and since this affects the nodes themselves, nextjs hydration fails (`Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.`) 

This error is both correct and incorrect- the server HTML does contain the `<template>`, but the browser removes it before hydration can happen. Is this a limitation with nextjs hydration? a bug in react-dom? can/should it be fixed? There's an open issue for react-dom here: https://github.com/facebook/react/issues/26071 but I think this will just remove the warning and not fix the uncaught hydration error. 

my workaround for now is to not use shadowrootmode and instead use the polyfill *without* removing the template item, just removing the contents:

```jsx
import { headers } from 'next/headers';

export async function ChromeSSR({ children }) {
  const nextHeaders = headers();
  const response = await fetch(api, {headers: nextHeaders});
  const html = await response.text();
  return (
    <div id="__chrome" style={{ display: 'contents' }} >
      <template
        suppressHydrationWarning
        dangerouslySetInnerHTML={{
          __html: html,
        }}
      />
      <script
        dangerouslySetInnerHTML={{
          __html: `
(function attachShadowRoots(root) {
  root.querySelectorAll("template").forEach(template => {
    const shadowRoot = template.parentNode.attachShadow({ mode: 'open' });
    shadowRoot.appendChild(template.content);
    template.innerHTML = '';
    attachShadowRoots(shadowRoot);
  });
})(document);
      `,
        }}
      />
      {children}
    </div>
  );
}

```

**Comment by augustjk** [comment]
I am interested in seeing a solution for this as well.
We're working on a SSR solution for web components built with Lit (https://github.com/lit/lit/tree/main/packages/labs/ssr-react).

Our solution so far has been to only render the `<template>` element containing the declarative shadow root in the server only. The package above does this by providing an alternative `jsx` or `createElement` function that will add the `<template>` element as a child, only when run in the server.

This has been working okay for the page router as the server-emitted `<template shadowrootmode="open">` elements would get consumed to attach the shadowroot during HTML parsing, and hydration used the client component definition which never produces the `<template>` element to begin with. Much like the approach described here: https://github.com/facebook/react/issues/26071#issuecomment-1712467891

However, this no longer works with the app router. It appears that the whole component tree as it was rendered in the server is serialized and sent to the client using a script tag with `__next_f.push()` at the bottom of the document. https://github.com/vercel/next.js/discussions/42170
This tree also includes the `<template>` element that isn't really needed for hydration in our case as the custom elements themselves have a separate mechanism for hydrating their (already attached by DSD) shadow root. It still leads to a critical hydration failure as the `<template>` element obviously is not found on the DOM.

This also causes problems when the page is constructed with Flight segments from the server when routing to a new page from the client. Since the page navigating to is server rendered, it includes the `<template>` element with a separate prop of `shadowrootmode`. This ends up creating an actual `<template>` element in the DOM tree as the browser is unable to treat this as declarative shadow DOM and attach the shadowroot as it should. This was not an issue for the page router I assume because it always deferred to client rendering in those cases.

The workaround for us is that we just can't have custom elements being server rendered with DSD within React Server Components.

**Reply by gaearon** [reply]
I'm curious if you tried applying this to the SSR pass in RSC. In RSC, SSR happens in a separate phase (https://github.com/reactwg/server-components/discussions/4). I think maybe the existing approach can work if you apply it *only* during the SSR pass (Client in RSC), but not during the RSC pass?

**Reply by gaearon** [reply]
In other words, if `exports` conditions has `react-server` (like [this](https://github.com/facebook/react/blob/269edb88b28b7394335fb504d9924acd35d4ef38/packages/react/package.json#L25)), then `createElement` shouldn't be hijacked. I wonder if that would work? Though I guess the later pass wouldn't use `createElement` as an API directly so maybe you'd have to intercept it at some other point.

**Reply by augustjk** [reply]
Indeed the existing approach should work with client component SSR. I was thinking we'd just have to document that limitation for users to make sure custom elements meant to be SSRed should stay behind the `'use client'` boundary.

I hadn't considered selectively patching `createElement` by the `react-server` condition. Though I think even if we do that and we don't SSR to produce declarative shadow DOM for the server tree, the behavior may very well be unexpected because custom element definitions may not get included with the client bundle so the user would just end up with a custom element that doesn't have its custom behavior/style etc.

So for now, any custom elements should be considered like React client components as they can have event bindings or internal state.

Down the line though, it would be nice if the server tree could handle declarative shadow DOM properly by not erroring on initial hydration, and actually attaching the shadowroot when parsing it so it can be useful for custom elements that don't have event bindings and don't need to be hydrated.

**Comment by michaelwarren1106** [comment]
+1 for Declarative Shadow DOM support in Next. SSR for custom elements is a major feature ask for lots of devs, and Declarative Shadow DOM is critical in enabling custom element SSR. NextJS could lead the pack with a first class solution.
